tests/components/huum/test_config_flow.py

```diff
@@ -0,0 +1,112 @@
+from homeassistant.components.huum.config_flow import CannotConnect
+from homeassistant.components.huum.const import DOMAIN
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import RESULT_TYPE_CREATE_ENTRY, RESULT_TYPE_FORM
+
+
+async def test_form(hass: HomeAssistant) -> None:
+    """Test we get the form."""
+    TEST_USERNAME = "test-username"
+
+    await setup.async_setup_component(hass, "persistent_notification", {})
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] is None
+
+    with patch(
+        "homeassistant.components.huum.config_flow.Huum.status",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.huum.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": TEST_USERNAME,
+                "password": "test-password",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == RESULT_TYPE_CREATE_ENTRY
+    assert result2["title"] == TEST_USERNAME
+    assert result2["data"] == {
+        "username": TEST_USERNAME,
+        "password": "test-password",
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.huum.config_flow.Huum.status",
+        side_effect=aiohttp.ClientError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == RESULT_TYPE_FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.huum.config_flow.Huum.status",
+        side_effect=CannotConnect,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == RESULT_TYPE_FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_unknown_error(hass: HomeAssistant) -> None:
```

Comment:
Checkout `@pytest.mark.parametrize`. you can make all these error tests parameterized so you don't have to have the same test logic thrice

Reply:
I agree that there are some redundant code here, but I personally like to keep tests separate like this to have better naming and better control over possible extra data that might be needed when possible future issues needs to be tested with regards to the same errors. Personal opinion I guess, hope that you are OK with me leaving it without the parameterization, but I can change if that is something that is required by the codebase.

Reply:
I mean, I am really a fan of keeping tests small and maintainable. With the requirement that config entry tests should end in create entry or abort, your tests will get big and parametrize is a perfect tool for this.

Keep in mind, design code on what you think is going to change. The only changes I usually see with config flows are 2 things: 1. a big overhaul where everything is changed 2. an extra exception (maybe your sauna needs to be on when you set it up to collect all data, and you found a way to catch that error in the lib and want to raise a separate error to give the user a better error message).

With the first case you probably always need to rewrite the tests, with the second one you just have to add an extra testcase and you are done.

This isn't a blocker, but I do have a strong preference